keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -9,105 +9,201 @@
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help kube-apiserver control the payload and thus the claims.
+
+## Design Details
+
+### ExternalJWTSigner RPC
+
+Will be served on a socket as configured via the `--key-service-url` flag.
 
 ```proto
 syntax = "proto3";
 
-package v1alpha1;
-
-service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+// This service is served by a process on a local Unix Domain Socket.
+service ExternalJWTSigner {
+  // Sign takes a serialized JWT payload, and returns the serialized header and
+  // signature.  The caller can then assemble the JWT from the header, payload,
+  // and signature.
+  //
+  // The plugin MUST set a key id in the returned JWT header.
+  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}
+
+  // FetchKeys returns the set of public keys that are trusted to sign
+  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
+message SignJWTRequest {
+  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
   bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message SignJWTResponse {
+  // The header for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  bytes header = 1;
+
+  // The signature for the JWT.  Already wrapped in URL-safe base64, valid for
+  // directly inserting into the final JWT.
+  bytes signature = 2;
 }
 
+message FetchKeysRequest {}
+
+message FetchKeysResponse {
+  repeated Key keys = 1;
+
+  // The timestamp when this data was pulled from the authoritative source of
+  // truth for verification keys.
+  google.protobuf.Timestamp data_timestamp = 2;
 
-message PublicKey {
-  // public_key is a PEM encoded public key
-  bytes public_key = 1;
-  // certificate is a concatenated list of PEM encoded x509 certificates
-  bytes certificates = 2;
-  // key_id is the key's ID
-  string key_id = 3;
-  // algorithm states the algorithm the key uses
-  string algorithm = 4;
+  // The timestamp for when it's appropriate to refetch verification keys to 
+  // pick changes if any.
+  google.protobuf.Timestamp suggested_next_fetch_at = 3;
```

Comment:
is there a minimum we'd accept here? (what would kube-apiserver do if this is before time.Now()? make sure we don't hot-loop or do something crazy if skew is encountered like that)

is this what informs the GetCacheAgeMaxSeconds() implementation for PublicKeysGetter? something like `suggested_next_fetch_at.Sub(time.Now())/time.Second`? that would mean the OIDC docs would return shrinking cache maxAge headers as we got closer and closer to the internal refresh time

Need to document how to use this and/or the metrics to safely add new keys and promote them to signing keys in an external signer

Reply:
the maxAge is eventually being used in cache control: https://github.com/kubernetes/kubernetes/blob/a15dca3765d1646d6c29832fdfe1a942e310f7a4/pkg/routes/openidmetadata.go#L95

if it slips to a negative value, it can cause a hot-loop. Makes me wonder if OIDCMetadata server shall at all depend on this field? If feels like a better option would be to have a fixed refresh interval for OIDC Metadata. Having said that, maybe we can omit out `suggested_next_fetch_at` (atleast to begin with)?

Reply:
what is the surface area of the JWK schema? I don't know offhand how sprawling / flexible the JWK schema is, and it seems like whatever could be returned from the external signer would would have to be plumbed all the way through kube-apiserver and published in oidc discovery... my goals for the initial iteration of this are to make the surface area of the issued tokens and published oidc discovery *identical* whether you're using an internal or external signer

Reply:
Done